Use .beam files after protocol consolidation

bblaszkow06 opened the pull request:
Hello! This PR is my proposal for solving #221.
Based on info from here https://github.com/elixir-lang/elixir/issues/7708 I've changed the args for dialyzer and instead of passing directories, I'm passing files list. The list is created in such a way that if files with the same name are present inside `dialyzer_paths()`, only the last one will be passed to the dialyzer. Combine that with consolidation path added by default after the compilation path and the issue seems solved.

A few notes, though:
* I haven't tested it with umbrella projects
* args list for dialyzer becomes enormous, I've added a limit of 10 for args inspection, but maybe some other approach should be taken (e.g. not printing the files at all or only with some flag)

asummers commented on lib/dialyxir/project.ex:
```diff
@@ -217,7 +226,7 @@ defmodule Dialyxir.Project do
   defp default_paths() do
     reduce_umbrella_children([], fn paths ->
-      [Mix.Project.compile_path() | paths]
+      [Mix.Project.compile_path() | [Mix.Project.consolidation_path() | paths]]
```
This can be done as:

```
[Mix.Project.compile_path(), Mix.Project.consolidation_path() | paths]
```

asummers commented on lib/dialyxir/project.ex:
```diff
@@ -52,6 +52,15 @@ defmodule Dialyxir.Project do
+    |> Enum.map(fn {_file, path} -> path |> to_charlist() end)
+  end
+
   def dialyzer_paths do
```
Is this called outside of this file anymore? Can probably be made private if not.

asummers commented:
Dialyzer is failing pretty spectacularly, and I'm uncomfortable merging this until it's tested on umbrellas.

asummers commented on lib/dialyxir/dialyzer.ex:
```diff
@@ -35,7 +35,7 @@ defmodule Dialyxir.Dialyzer do
 
         args
-        |> inspect(label: "dialyzer args", pretty: true)
+        |> inspect(label: "dialyzer args", pretty: true, limit: 10)
```
I wonder if this should be an opt of some sort instead

bblaszkow06 replied:
I believe there could be a `--debug`/`--verbose` that would print the files (`limit: :infinity`) and by default, the files wouldn't be printed at all. Anyway, I've reverted that change, I'll leave the decision up to you

asummers replied:
Exactly my thought. Adding that opt is probably out of scope of this anyway, but good call on bringing that up.

bblaszkow06 commented:
@asummers I've tested an umbrella project and it turned out that consolidation path can be included just once (as it's shared for the whole umbrella project).

The 'spectacular failure' is actually not a bug, it's a feature as those are valid `:underspec` warnings :stuck_out_tongue: 
It seems that this has been fixed in Elixir 1.8 (here I believe: https://github.com/elixir-lang/elixir/pull/7832).

I've added Elixir 1.8.1 to the build matrix on Travis (and fixed the exclusion rule by the way) and as you can see (https://travis-ci.org/jeremyjh/dialyxir/builds/498428066), the build succeeds for this version. I'm not quite sure what to do with the failing versions - should they be removed? Or 'fixed' by silencing the warnings in `.dialyzer_ignore.exs`?

jeremyjh commented:
I think we're going to have to do something about the output including the entire files collection, this will be many pages of useless spam in larger projects.

Also, we shouldn't be getting warnings from Elixir lib files in our results should we? 
For example one of the warnings is in this file: `/home/build/elixir/lib/elixir/lib/list/chars.ex:1:contract_supertype`

I think its because this file is included: 
   '/home/travis/build/jeremyjh/dialyxir/_build/prod/lib/dialyxir/consolidated/Elixir.String.Chars.beam',

jeremyjh commented:
@bblaszkow06 are you planning to finish this soon? I'd like to get into 1.0 and to me it seems like it is pretty close. Let me know if you need help.

bblaszkow06 commented:
@jeremyjh Sorry, I haven't had much time to finish this, but I managed to find a way to prevent seeing errors from Elixir lib files by taking from consolidation path only the files that were present in compilation path(s).
Now I have to clean up the code I was experimenting on and I hope I will find time for that by the end of this week.

bblaszkow06 commented:
@jeremyjh All green and ready for review. :smile: There are some warnings printed for `MIX_ENV=prod mix dialyzer --format raw` test, but they are present on master as well

jeremyjh commented:
Looks good to me, thank you!
